File: apps/langgraph-worker/app/utils/api_errors.py
"""
API Error Classification Module for LangGraph Worker

Provides standardized API error classification that matches the TypeScript
types in @genni/shared-types for consistent error handling across the platform.
"""

from dataclasses import dataclass, asdict
from enum import Enum
from typing import Optional, Dict, Any
import time


class ApiProvider(str, Enum):
    """API provider identifiers matching TypeScript ApiProvider type"""
    GOOGLE_PLACES = "google_places"
    FINDYMAIL = "findymail"
    PERPLEXITY = "perplexity"
    TAVILY = "tavily"
    OPENAI = "openai"
    FASTSPRING = "fastspring"
    UNKNOWN = "unknown"


class ApiErrorCategory(str, Enum):
    """Error categories matching TypeScript ApiErrorCategory type"""
    AUTHENTICATION = "authentication"
    AUTHORIZATION = "authorization"
    QUOTA_EXHAUSTED = "quota_exhausted"
    RATE_LIMITED = "rate_limited"
    RATE_LIMIT_EXCEEDED = "rate_limit_exceeded"
    INVALID_REQUEST = "invalid_request"
    NOT_FOUND = "not_found"
    SERVER_ERROR = "server_error"
    TIMEOUT = "timeout"
    NETWORK_ERROR = "network_error"
    UNKNOWN = "unknown"


class ErrorSeverity(str, Enum):
    """Error severity levels matching TypeScript ErrorSeverity type"""
    INFO = "info"
    WARNING = "warning"
    ERROR = "error"
    CRITICAL = "critical"


class ErrorAction(str, Enum):
    """Suggested actions matching TypeScript ErrorAction type"""
    RETRY = "retry"
    WAIT = "wait"
    CHECK_API_KEY = "check_api_key"
    ADD_CREDITS = "add_credits"
    UPGRADE_PLAN = "upgrade_plan"
    CHECK_BILLING = "check_billing"
    CONTACT_SUPPORT = "contact_support"
    NONE = "none"


@dataclass
class StandardizedApiError:
    """
    Standardized API error matching TypeScript ApiError interface.
    Used for consistent error handling across Python and TypeScript codebases.
    """
    error_code: str
    provider: ApiProvider
    category: ApiErrorCategory
    user_message: str
    severity: ErrorSeverity
    retryable: bool
    suggested_action: ErrorAction
    technical_message: Optional[str] = None
    action_label: Optional[str] = None
    action_url: Optional[str] = None
    retry_after_ms: Optional[int] = None
    original_status: Optional[int] = None
    original_message: Optional[str] = None
    timestamp: Optional[int] = None

    def __post_init__(self):
        if self.timestamp is None:
            self.timestamp = int(time.time() * 1000)

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary for JSON serialization and API responses"""
        result = {
            "errorCode": self.error_code,
            "provider": self.provider.value if isinstance(self.provider, ApiProvider) else self.provider,
            "category": self.category.value if isinstance(self.category, ApiErrorCategory) else self.category,
            "userMessage": self.user_message,
            "severity": self.severity.value if isinstance(self.severity, ErrorSeverity) else self.severity,
            "retryable": self.retryable,
            "suggestedAction": self.suggested_action.value if isinstance(self.suggested_action, ErrorAction) else self.suggested_action,
            "timestamp": self.timestamp,
        }

        if self.technical_message:
            result["technicalMessage"] = self.technical_message
        if self.action_label:
            result["actionLabel"] = self.action_label
        if self.action_url:
            result["actionUrl"] = self.action_url
        if self.retry_after_ms:
            result["retryAfterMs"] = self.retry_after_ms
        if self.original_status:
            result["originalStatus"] = self.original_status
        if self.original_message:
            result["originalMessage"] = self.original_message

        return result


# ============================================
# API Error Codes (matching TypeScript)
# ============================================

class API_ERROR_CODES:
    """API error codes matching TypeScript API_ERROR_CODES"""
    # Google Places
    GOOGLE_QUOTA_EXHAUSTED = "GOOGLE_QUOTA_EXHAUSTED"
    GOOGLE_AUTH_FAILED = "GOOGLE_AUTH_FAILED"
    GOOGLE_REQUEST_DENIED = "GOOGLE_REQUEST_DENIED"
    GOOGLE_INVALID_REQUEST = "GOOGLE_INVALID_REQUEST"
    GOOGLE_ZERO_RESULTS = "GOOGLE_ZERO_RESULTS"
    GOOGLE_SERVER_ERROR = "GOOGLE_SERVER_ERROR"

    # FindyMail
    FINDYMAIL_AUTH_FAILED = "FINDYMAIL_AUTH_FAILED"
    FINDYMAIL_CREDITS_EXHAUSTED = "FINDYMAIL_CREDITS_EXHAUSTED"
    FINDYMAIL_RATE_LIMITED = "FINDYMAIL_RATE_LIMITED"
    FINDYMAIL_INVALID_REQUEST = "FINDYMAIL_INVALID_REQUEST"
    FINDYMAIL_SERVER_ERROR = "FINDYMAIL_SERVER_ERROR"

    # Perplexity
    PERPLEXITY_AUTH_FAILED = "PERPLEXITY_AUTH_FAILED"
    PERPLEXITY_CREDITS_EXHAUSTED = "PERPLEXITY_CREDITS_EXHAUSTED"
    PERPLEXITY_RATE_LIMITED = "PERPLEXITY_RATE_LIMITED"
    PERPLEXITY_RATE_EXCEEDED = "PERPLEXITY_RATE_EXCEEDED"
    PERPLEXITY_TIMEOUT = "PERPLEXITY_TIMEOUT"
    PERPLEXITY_SERVER_ERROR = "PERPLEXITY_SERVER_ERROR"

    # OpenAI
    OPENAI_AUTH_FAILED = "OPENAI_AUTH_FAILED"
    OPENAI_QUOTA_EXHAUSTED = "OPENAI_QUOTA_EXHAUSTED"
    OPENAI_RATE_LIMITED = "OPENAI_RATE_LIMITED"
    OPENAI_CONTEXT_LENGTH = "OPENAI_CONTEXT_LENGTH"
    OPENAI_SERVER_ERROR = "OPENAI_SERVER_ERROR"

    # Tavily
    TAVILY_AUTH_FAILED = "TAVILY_AUTH_FAILED"
    TAVILY_CREDITS_EXHAUSTED = "TAVILY_CREDITS_EXHAUSTED"
    TAVILY_RATE_LIMITED = "TAVILY_RATE_LIMITED"
    TAVILY_SERVER_ERROR = "TAVILY_SERVER_ERROR"

    # Generic
    UNKNOWN_ERROR = "UNKNOWN_ERROR"
    NETWORK_ERROR = "NETWORK_ERROR"
    TIMEOUT_ERROR = "TIMEOUT_ERROR"


# ============================================
# Error Messages (matching TypeScript)
# ============================================

ERROR_MESSAGES: Dict[str, Dict[str, Any]] = {
    # Perplexity
    API_ERROR_CODES.PERPLEXITY_AUTH_FAILED: {
        "user_message": "Perplexity API key is invalid or expired. Please update your API key in Settings.",
        "severity": ErrorSeverity.ERROR,
        "retryable": False,
        "suggested_action": ErrorAction.CHECK_API_KEY,
        "action_label": "Update API Key",
        "action_url": "/settings/api-keys",
    },
    API_ERROR_CODES.PERPLEXITY_CREDITS_EXHAUSTED: {
        "user_message": "Perplexity billing limit reached. Please check your Perplexity account.",
        "severity": ErrorSeverity.ERROR,
        "retryable": False,
        "suggested_action": ErrorAction.CHECK_BILLING,
        "action_label": "Check Billing",
        "action_url": "https://www.perplexity.ai/settings/api",
    },
    API_ERROR_CODES.PERPLEXITY_RATE_LIMITED: {
        "user_message": "Perplexity is temporarily rate limited. Retrying automatically...",
        "severity": ErrorSeverity.WARNING,
        "retryable": True,
        "suggested_action": ErrorAction.WAIT,
    },
    API_ERROR_CODES.PERPLEXITY_RATE_EXCEEDED: {
        "user_message": "Perplexity rate limit exceeded. Research quality may be reduced.",
        "severity": ErrorSeverity.WARNING,
        "retryable": False,
        "suggested_action": ErrorAction.UPGRADE_PLAN,
        "action_label": "Upgrade Plan",
    },
    API_ERROR_CODES.PERPLEXITY_TIMEOUT: {
        "user_message": "Perplexity request timed out. Retrying...",
        "severity": ErrorSeverity.WARNING,
        "retryable": True,
        "suggested_action": ErrorAction.RETRY,
    },
    API_ERROR_CODES.PERPLEXITY_SERVER_ERROR: {
        "user_message": "Perplexity service is temporarily unavailable. Retrying...",
        "severity": ErrorSeverity.WARNING,
        "retryable": True,
        "suggested_action": ErrorAction.RETRY,
    },

    # OpenAI
    API_ERROR_CODES.OPENAI_AUTH_FAILED: {
        "user_message": "OpenAI API key is invalid or expired. Please update your API key in Settings.",
        "severity": ErrorSeverity.ERROR,
        "retryable": False,
        "suggested_action": ErrorAction.CHECK_API_KEY,
        "action_label": "Update API Key",
        "action_url": "/settings/api-keys",
    },
    API_ERROR_CODES.OPENAI_QUOTA_EXHAUSTED: {
        "user_message": "OpenAI billing limit reached. Please check your OpenAI account billing.",
        "severity": ErrorSeverity.ERROR,
        "retryable": False,
        "suggested_action": ErrorAction.CHECK_BILLING,
        "action_label": "Check Billing",
        "action_url": "https://platform.openai.com/account/billing",
    },
    API_ERROR_CODES.OPENAI_RATE_LIMITED: {
        "user_message": "OpenAI is temporarily rate limited. Retrying automatically...",
        "severity": ErrorSeverity.WARNING,
        "retryable": True,
        "suggested_action": ErrorAction.WAIT,
    },
    API_ERROR_CODES.OPENAI_CONTEXT_LENGTH: {
        "user_message": "Request too large for OpenAI model. Reducing input size...",
        "severity": ErrorSeverity.WARNING,
        "retryable": False,
        "suggested_action": ErrorAction.NONE,
    },
    API_ERROR_CODES.OPENAI_SERVER_ERROR: {
        "user_message": "OpenAI service is temporarily unavailable. Retrying...",
        "severity": ErrorSeverity.WARNING,
        "retryable": True,
        "suggested_action": ErrorAction.RETRY,
    },

    # Generic
    API_ERROR_CODES.UNKNOWN_ERROR: {
        "user_message": "An unexpected error occurred. Please try again.",
        "severity": ErrorSeverity.ERROR,
        "retryable": True,
        "suggested_action": ErrorAction.RETRY,
    },
    API_ERROR_CODES.NETWORK_ERROR: {
        "user_message": "Network connection error. Please check your internet connection.",
        "severity": ErrorSeverity.WARNING,
        "retryable": True,
        "suggested_action": ErrorAction.RETRY,
    },
    API_ERROR_CODES.TIMEOUT_ERROR: {
        "user_message": "Request timed out. Retrying...",
        "severity": ErrorSeverity.WARNING,
        "retryable": True,
        "suggested_action": ErrorAction.RETRY,
    },
}


def get_error_message(error_code: str) -> Dict[str, Any]:
    """Get error message configuration for an error code"""
    return ERROR_MESSAGES.get(error_code, ERROR_MESSAGES[API_ERROR_CODES.UNKNOWN_ERROR])


# ============================================
# Perplexity Error Classification
# ============================================

def classify_perplexity_error(
    http_status: int,
    error_message: str = "",
    retry_after: Optional[int] = None
) -> StandardizedApiError:
    """
    Classify Perplexity API errors into standardized format.

    Args:
        http_status: HTTP status code from Perplexity API
        error_message: Error message from response
        retry_after: Optional Retry-After header value in seconds

    Returns:
        StandardizedApiError with appropriate classification
    """
    base = {
        "provider": ApiProvider.PERPLEXITY,
        "original_status": http_status,
        "original_message": error_message,
    }

    if http_status == 401:
        msg = get_error_message(API_ERROR_CODES.PERPLEXITY_AUTH_FAILED)
        return StandardizedApiError(
            error_code=API_ERROR_CODES.PERPLEXITY_AUTH_FAILED,
            category=ApiErrorCategory.AUTHENTICATION,
            **base,
            **msg
        )

    if http_status == 402:
        msg = get_error_message(API_ERROR_CODES.PERPLEXITY_CREDITS_EXHAUSTED)
        return StandardizedApiError(
            error_code=API_ERROR_CODES.PERPLEXITY_CREDITS_EXHAUSTED,
            category=ApiErrorCategory.QUOTA_EXHAUSTED,
            **base,
            **msg
        )

    if http_status == 429:
        # Check if it's quota vs transient rate limiting
        error_lower = error_message.lower()
        is_quota_exceeded = any(term in error_lower for term in ["quota", "billing", "limit exceeded"])

        if is_quota_exceeded:
            msg = get_error_message(API_ERROR_CODES.PERPLEXITY_RATE_EXCEEDED)
            return StandardizedApiError(
                error_code=API_ERROR_CODES.PERPLEXITY_RATE_EXCEEDED,
                category=ApiErrorCategory.RATE_LIMIT_EXCEEDED,
                **base,
                **msg
            )

        msg = get_error_message(API_ERROR_CODES.PERPLEXITY_RATE_LIMITED)
        return StandardizedApiError(
            error_code=API_ERROR_CODES.PERPLEXITY_RATE_LIMITED,
            category=ApiErrorCategory.RATE_LIMITED,
            retry_after_ms=retry_after * 1000 if retry_after else 5000,
            **base,
            **msg
        )

    if http_status == 408:
        msg = get_error_message(API_ERROR_CODES.PERPLEXITY_TIMEOUT)
        return StandardizedApiError(
            error_code=API_ERROR_CODES.PERPLEXITY_TIMEOUT,
            category=ApiErrorCategory.TIMEOUT,
            retry_after_ms=2000,
            **base,
            **msg
        )

    if http_status >= 500:
        msg = get_error_message(API_ERROR_CODES.PERPLEXITY_SERVER_ERROR)
        return StandardizedApiError(
            error_code=API_ERROR_CODES.PERPLEXITY_SERVER_ERROR,
            category=ApiErrorCategory.SERVER_ERROR,
            retry_after_ms=5000,
            **base,
            **msg
        )

    # Unknown error
    msg = get_error_message(API_ERROR_CODES.UNKNOWN_ERROR)
    return StandardizedApiError(
        error_code=API_ERROR_CODES.UNKNOWN_ERROR,
        category=ApiErrorCategory.SERVER_ERROR if http_status >= 500 else ApiErrorCategory.UNKNOWN,
        technical_message=error_message,
        **base,
        **msg
    )


# ============================================
# OpenAI Error Classification
# ============================================

def classify_openai_error(
    http_status: int,
    error_message: str = "",
    error_type: Optional[str] = None
) -> StandardizedApiError:
    """
    Classify OpenAI API errors into standardized format.

    Args:
        http_status: HTTP status code from OpenAI API
        error_message: Error message from response
        error_type: OpenAI error type (e.g., "insufficient_quota")

    Returns:
        StandardizedApiError with appropriate classification
    """
    base = {
        "provider": ApiProvider.OPENAI,
        "original_status": http_status,
        "original_message": error_message,
    }

    error_lower = error_message.lower()

    # Check for quota/billing issues
    if error_type == "insufficient_quota" or "quota" in error_lower or "billing" in error_lower:
        msg = get_error_message(API_ERROR_CODES.OPENAI_QUOTA_EXHAUSTED)
        return StandardizedApiError(
            error_code=API_ERROR_CODES.OPENAI_QUOTA_EXHAUSTED,
            category=ApiErrorCategory.QUOTA_EXHAUSTED,
            **base,
            **msg
        )

    # Context length exceeded
    if "context length" in error_lower:
        msg = get_error_message(API_ERROR_CODES.OPENAI_CONTEXT_LENGTH)
        return StandardizedApiError(
            error_code=API_ERROR_CODES.OPENAI_CONTEXT_LENGTH,
            category=ApiErrorCategory.INVALID_REQUEST,
            **base,
            **msg
        )

    if http_status == 401:
        msg = get_error_message(API_ERROR_CODES.OPENAI_AUTH_FAILED)
        return StandardizedApiError(
            error_code=API_ERROR_CODES.OPENAI_AUTH_FAILED,
            category=ApiErrorCategory.AUTHENTICATION,
            **base,
            **msg
        )

    if http_status == 429:
        msg = get_error_message(API_ERROR_CODES.OPENAI_RATE_LIMITED)
        return StandardizedApiError(
            error_code=API_ERROR_CODES.OPENAI_RATE_LIMITED,
            category=ApiErrorCategory.RATE_LIMITED,
            retry_after_ms=10000,  # OpenAI rate limits can be longer
            **base,
            **msg
        )

    if http_status >= 500:
        msg = get_error_message(API_ERROR_CODES.OPENAI_SERVER_ERROR)
        return StandardizedApiError(
            error_code=API_ERROR_CODES.OPENAI_SERVER_ERROR,
            category=ApiErrorCategory.SERVER_ERROR,
            retry_after_ms=5000,
            **base,
            **msg
        )

    # Unknown error
    msg = get_error_message(API_ERROR_CODES.UNKNOWN_ERROR)
    return StandardizedApiError(
        error_code=API_ERROR_CODES.UNKNOWN_ERROR,
        category=ApiErrorCategory.SERVER_ERROR if http_status >= 500 else ApiErrorCategory.UNKNOWN,
        technical_message=error_message,
        **base,
        **msg
    )


# ============================================
# Helper Functions
# ============================================

def should_block_pipeline(api_error: StandardizedApiError) -> bool:
    """
    Check if an API error should block the pipeline.
    Matches the TypeScript shouldBlockPipeline function.

    User-actionable errors like auth failed or credits exhausted
    should stop the pipeline immediately with a clear message.
    """
    blocking_categories = [
        ApiErrorCategory.AUTHENTICATION,
        ApiErrorCategory.AUTHORIZATION,
        ApiErrorCategory.QUOTA_EXHAUSTED,
        ApiErrorCategory.RATE_LIMIT_EXCEEDED,
    ]
    return api_error.category in blocking_categories


def is_user_actionable_error(api_error: StandardizedApiError) -> bool:
    """
    Check if an error requires user action (not just wait/retry).
    Matches the TypeScript isUserActionableApiError function.
    """
    actionable_categories = [
        ApiErrorCategory.AUTHENTICATION,
        ApiErrorCategory.AUTHORIZATION,
        ApiErrorCategory.QUOTA_EXHAUSTED,
        ApiErrorCategory.RATE_LIMIT_EXCEEDED,
    ]
    return api_error.category in actionable_categories


class ApiErrorException(Exception):
    """
    Exception wrapper for StandardizedApiError.
    Use this to throw errors that should be surfaced to users.
    """
    def __init__(self, api_error: StandardizedApiError):
        self.api_error = api_error
        super().__init__(api_error.user_message)

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary for API responses"""
        return self.api_error.to_dict()

File: apps/convex-backend/convex/lib/apiErrors.ts
/**
 * API Error Classification Module
 *
 * Classifies external API errors into standardized ApiError format
 * for consistent handling across the backend and frontend.
 */

import { ConvexError } from "convex/values";
import type {
  ApiError,
  ApiProvider,
  ApiErrorCategory,
  ErrorSeverity,
  ErrorAction,
  ApiErrorCode,
} from "@genni/shared-types";
import {
  API_ERROR_CODES,
  ERROR_MESSAGES,
  getErrorMessage,
} from "@genni/shared-types";

// ============================================
// Google Maps/Places Error Classification
// ============================================

/**
 * Classify Google Maps/Places API errors
 * @param status - Places API status string (OK, ZERO_RESULTS, OVER_QUERY_LIMIT, etc.)
 * @param errorMessage - Optional error message from response
 * @param httpStatus - Optional HTTP status code
 */
export function classifyGoogleError(
  status: string,
  errorMessage?: string,
  httpStatus?: number
): ApiError {
  const base = {
    provider: "google_places" as ApiProvider,
    timestamp: Date.now(),
    originalStatus: httpStatus,
    originalMessage: errorMessage,
  };

  switch (status) {
    case "OVER_QUERY_LIMIT":
    case "RESOURCE_EXHAUSTED":
      return {
        ...base,
        errorCode: API_ERROR_CODES.GOOGLE_QUOTA_EXHAUSTED,
        category: "quota_exhausted",
        ...getErrorMessage(API_ERROR_CODES.GOOGLE_QUOTA_EXHAUSTED),
      };

    case "REQUEST_DENIED":
      // Check if it's an auth error or permission error
      const isAuthError =
        errorMessage?.toLowerCase().includes("api key") ||
        errorMessage?.toLowerCase().includes("invalid key");

      if (isAuthError) {
        return {
          ...base,
          errorCode: API_ERROR_CODES.GOOGLE_AUTH_FAILED,
          category: "authentication",
          ...getErrorMessage(API_ERROR_CODES.GOOGLE_AUTH_FAILED),
        };
      }

      return {
        ...base,
        errorCode: API_ERROR_CODES.GOOGLE_REQUEST_DENIED,
        category: "authorization",
        userMessage:
          "Google Maps API request was denied. Please check your API key permissions.",
        severity: "error" as ErrorSeverity,
        retryable: false,
        suggestedAction: "check_api_key" as ErrorAction,
        actionLabel: "Check API Key",
        actionUrl: "/settings/api-keys",
      };

    case "INVALID_REQUEST":
      return {
        ...base,
        errorCode: API_ERROR_CODES.GOOGLE_INVALID_REQUEST,
        category: "invalid_request",
        ...getErrorMessage(API_ERROR_CODES.GOOGLE_INVALID_REQUEST),
      };

    case "ZERO_RESULTS":
      return {
        ...base,
        errorCode: API_ERROR_CODES.GOOGLE_ZERO_RESULTS,
        category: "not_found",
        ...getErrorMessage(API_ERROR_CODES.GOOGLE_ZERO_RESULTS),
      };

    case "UNKNOWN_ERROR":
    default:
      // Check HTTP status for server errors
      if (httpStatus && httpStatus >= 500) {
        return {
          ...base,
          errorCode: API_ERROR_CODES.GOOGLE_SERVER_ERROR,
          category: "server_error",
          ...getErrorMessage(API_ERROR_CODES.GOOGLE_SERVER_ERROR),
        };
      }

      return {
        ...base,
        errorCode: API_ERROR_CODES.UNKNOWN_ERROR,
        category: "unknown",
        ...getErrorMessage(API_ERROR_CODES.UNKNOWN_ERROR),
        technicalMessage: errorMessage,
      };
  }
}

// ============================================
// FindyMail Error Classification
// ============================================

/**
 * Classify FindyMail API errors
 * @param httpStatus - HTTP status code
 * @param errorBody - Response body or error message
 */
export function classifyFindyMailError(
  httpStatus: number,
  errorBody?: any
): ApiError {
  const errorMessage =
    typeof errorBody === "string"
      ? errorBody
      : errorBody?.message || errorBody?.error || "";

  const base = {
    provider: "findymail" as ApiProvider,
    timestamp: Date.now(),
    originalStatus: httpStatus,
    originalMessage: errorMessage,
  };

  switch (httpStatus) {
    case 401:
      return {
        ...base,
        errorCode: API_ERROR_CODES.FINDYMAIL_AUTH_FAILED,
        category: "authentication",
        ...getErrorMessage(API_ERROR_CODES.FINDYMAIL_AUTH_FAILED),
      };

    case 402:
      // Payment required = credits exhausted
      return {
        ...base,
        errorCode: API_ERROR_CODES.FINDYMAIL_CREDITS_EXHAUSTED,
        category: "quota_exhausted",
        ...getErrorMessage(API_ERROR_CODES.FINDYMAIL_CREDITS_EXHAUSTED),
      };

    case 423:
      // Subscription is paused (per FindyMail API docs)
      return {
        ...base,
        errorCode: API_ERROR_CODES.FINDYMAIL_SUBSCRIPTION_PAUSED,
        category: "subscription_paused",
        ...getErrorMessage(API_ERROR_CODES.FINDYMAIL_SUBSCRIPTION_PAUSED),
      };

    case 429:
      // Check if it's transient rate limiting or quota exhaustion
      const isQuotaExhausted =
        errorMessage.toLowerCase().includes("quota") ||
        errorMessage.toLowerCase().includes("limit exceeded") ||
        errorMessage.toLowerCase().includes("credits");

      if (isQuotaExhausted) {
        return {
          ...base,
          errorCode: API_ERROR_CODES.FINDYMAIL_CREDITS_EXHAUSTED,
          category: "rate_limit_exceeded",
          ...getErrorMessage(API_ERROR_CODES.FINDYMAIL_CREDITS_EXHAUSTED),
        };
      }

      // Transient rate limiting
      const retryAfterMs = parseRetryAfterHeader(errorBody);
      return {
        ...base,
        errorCode: API_ERROR_CODES.FINDYMAIL_RATE_LIMITED,
        category: "rate_limited",
        ...getErrorMessage(API_ERROR_CODES.FINDYMAIL_RATE_LIMITED),
        retryAfterMs,
      };

    case 400:
      return {
        ...base,
        errorCode: API_ERROR_CODES.FINDYMAIL_INVALID_REQUEST,
        category: "invalid_request",
        ...getErrorMessage(API_ERROR_CODES.FINDYMAIL_INVALID_REQUEST),
        technicalMessage: errorMessage,
      };

    case 500:
    case 502:
    case 503:
    case 504:
      return {
        ...base,
        errorCode: API_ERROR_CODES.FINDYMAIL_SERVER_ERROR,
        category: "server_error",
        ...getErrorMessage(API_ERROR_CODES.FINDYMAIL_SERVER_ERROR),
        retryAfterMs: 5000,
      };

    default:
      return {
        ...base,
        errorCode: API_ERROR_CODES.UNKNOWN_ERROR,
        category: httpStatus >= 500 ? "server_error" : "unknown",
        ...getErrorMessage(API_ERROR_CODES.UNKNOWN_ERROR),
        retryable: httpStatus >= 500,
        technicalMessage: errorMessage,
      };
  }
}

/**
 * Create a "no contacts found" result for FindyMail
 * This is NOT an error - it's a successful API call with no results
 */
export function createFindyMailNoContactsResult(): ApiError {
  return {
    provider: "findymail" as ApiProvider,
    timestamp: Date.now(),
    errorCode: API_ERROR_CODES.FINDYMAIL_NO_CONTACTS_FOUND,
    category: "no_results",
    ...getErrorMessage(API_ERROR_CODES.FINDYMAIL_NO_CONTACTS_FOUND),
  };
}

// ============================================
// Perplexity Error Classification
// ============================================

/**
 * Classify Perplexity API errors
 * @param httpStatus - HTTP status code
 * @param errorMessage - Error message
 * @param retryAfter - Optional Retry-After header value in seconds
 */
export function classifyPerplexityError(
  httpStatus: number,
  errorMessage: string = "",
  retryAfter?: number
): ApiError {
  const base = {
    provider: "perplexity" as ApiProvider,
    timestamp: Date.now(),
    originalStatus: httpStatus,
    originalMessage: errorMessage,
  };

  switch (httpStatus) {
    case 401:
      return {
        ...base,
        errorCode: API_ERROR_CODES.PERPLEXITY_AUTH_FAILED,
        category: "authentication",
        ...getErrorMessage(API_ERROR_CODES.PERPLEXITY_AUTH_FAILED),
      };

    case 402:
      return {
        ...base,
        errorCode: API_ERROR_CODES.PERPLEXITY_CREDITS_EXHAUSTED,
        category: "quota_exhausted",
        ...getErrorMessage(API_ERROR_CODES.PERPLEXITY_CREDITS_EXHAUSTED),
      };

    case 429:
      // Check if it's quota vs transient rate limiting
      const isQuotaExceeded =
        errorMessage.toLowerCase().includes("quota") ||
        errorMessage.toLowerCase().includes("billing") ||
        errorMessage.toLowerCase().includes("limit exceeded");

      if (isQuotaExceeded) {
        return {
          ...base,
          errorCode: API_ERROR_CODES.PERPLEXITY_RATE_EXCEEDED,
          category: "rate_limit_exceeded",
          ...getErrorMessage(API_ERROR_CODES.PERPLEXITY_RATE_EXCEEDED),
        };
      }

      return {
        ...base,
        errorCode: API_ERROR_CODES.PERPLEXITY_RATE_LIMITED,
        category: "rate_limited",
        ...getErrorMessage(API_ERROR_CODES.PERPLEXITY_RATE_LIMITED),
        retryAfterMs: retryAfter ? retryAfter * 1000 : 5000,
      };

    case 408:
      return {
        ...base,
        errorCode: API_ERROR_CODES.PERPLEXITY_TIMEOUT,
        category: "timeout",
        ...getErrorMessage(API_ERROR_CODES.PERPLEXITY_TIMEOUT),
        retryAfterMs: 2000,
      };

    case 500:
    case 502:
    case 503:
    case 504:
      return {
        ...base,
        errorCode: API_ERROR_CODES.PERPLEXITY_SERVER_ERROR,
        category: "server_error",
        ...getErrorMessage(API_ERROR_CODES.PERPLEXITY_SERVER_ERROR),
        retryAfterMs: 5000,
      };

    default:
      return {
        ...base,
        errorCode: API_ERROR_CODES.UNKNOWN_ERROR,
        category: httpStatus >= 500 ? "server_error" : "unknown",
        ...getErrorMessage(API_ERROR_CODES.UNKNOWN_ERROR),
        retryable: httpStatus >= 500,
        technicalMessage: errorMessage,
      };
  }
}

// ============================================
// OpenAI Error Classification
// ============================================

/**
 * Classify OpenAI API errors
 * @param httpStatus - HTTP status code
 * @param errorMessage - Error message
 * @param errorType - OpenAI error type (e.g., "insufficient_quota")
 */
export function classifyOpenAIError(
  httpStatus: number,
  errorMessage: string = "",
  errorType?: string
): ApiError {
  const base = {
    provider: "openai" as ApiProvider,
    timestamp: Date.now(),
    originalStatus: httpStatus,
    originalMessage: errorMessage,
  };

  // Check for specific error types
  if (
    errorType === "insufficient_quota" ||
    errorMessage.toLowerCase().includes("quota") ||
    errorMessage.toLowerCase().includes("billing")
  ) {
    return {
      ...base,
      errorCode: API_ERROR_CODES.OPENAI_QUOTA_EXHAUSTED,
      category: "quota_exhausted",
      ...getErrorMessage(API_ERROR_CODES.OPENAI_QUOTA_EXHAUSTED),
    };
  }

  if (errorMessage.toLowerCase().includes("context length")) {
    return {
      ...base,
      errorCode: API_ERROR_CODES.OPENAI_CONTEXT_LENGTH,
      category: "invalid_request",
      ...getErrorMessage(API_ERROR_CODES.OPENAI_CONTEXT_LENGTH),
    };
  }

  switch (httpStatus) {
    case 401:
      return {
        ...base,
        errorCode: API_ERROR_CODES.OPENAI_AUTH_FAILED,
        category: "authentication",
        ...getErrorMessage(API_ERROR_CODES.OPENAI_AUTH_FAILED),
      };

    case 429:
      return {
        ...base,
        errorCode: API_ERROR_CODES.OPENAI_RATE_LIMITED,
        category: "rate_limited",
        ...getErrorMessage(API_ERROR_CODES.OPENAI_RATE_LIMITED),
        retryAfterMs: 10000, // OpenAI rate limits can be longer
      };

    case 500:
    case 502:
    case 503:
    case 504:
      return {
        ...base,
        errorCode: API_ERROR_CODES.OPENAI_SERVER_ERROR,
        category: "server_error",
        ...getErrorMessage(API_ERROR_CODES.OPENAI_SERVER_ERROR),
        retryAfterMs: 5000,
      };

    default:
      return {
        ...base,
        errorCode: API_ERROR_CODES.UNKNOWN_ERROR,
        category: httpStatus >= 500 ? "server_error" : "unknown",
        ...getErrorMessage(API_ERROR_CODES.UNKNOWN_ERROR),
        retryable: httpStatus >= 500,
        technicalMessage: errorMessage,
      };
  }
}

// ============================================
// FastSpring (Billing) Error Classification
// ============================================

/**
 * Classify FastSpring billing API errors
 * @param httpStatus - HTTP status code
 * @param errorMessage - Error message
 */
export function classifyFastSpringError(
  httpStatus: number,
  errorMessage: string = ""
): ApiError {
  const base = {
    provider: "fastspring" as ApiProvider,
    timestamp: Date.now(),
    originalStatus: httpStatus,
    originalMessage: errorMessage,
  };

  switch (httpStatus) {
    case 401:
    case 403:
      return {
        ...base,
        errorCode: API_ERROR_CODES.FASTSPRING_AUTH_FAILED,
        category: "authentication",
        ...getErrorMessage(API_ERROR_CODES.FASTSPRING_AUTH_FAILED),
      };

    case 402:
      return {
        ...base,
        errorCode: API_ERROR_CODES.FASTSPRING_PAYMENT_FAILED,
        category: "quota_exhausted",
        ...getErrorMessage(API_ERROR_CODES.FASTSPRING_PAYMENT_FAILED),
      };

    case 404:
      return {
        ...base,
        errorCode: API_ERROR_CODES.FASTSPRING_ORDER_NOT_FOUND,
        category: "not_found",
        ...getErrorMessage(API_ERROR_CODES.FASTSPRING_ORDER_NOT_FOUND),
      };

    case 500:
    case 502:
    case 503:
    case 504:
      return {
        ...base,
        errorCode: API_ERROR_CODES.FASTSPRING_SERVER_ERROR,
        category: "server_error",
        ...getErrorMessage(API_ERROR_CODES.FASTSPRING_SERVER_ERROR),
        retryAfterMs: 5000,
      };

    default:
      return {
        ...base,
        errorCode: API_ERROR_CODES.UNKNOWN_ERROR,
        category: httpStatus >= 500 ? "server_error" : "unknown",
        ...getErrorMessage(API_ERROR_CODES.UNKNOWN_ERROR),
        retryable: httpStatus >= 500,
        technicalMessage: errorMessage,
      };
  }
}

// ============================================
// Helper Functions
// ============================================

/**
 * Parse Retry-After header from response
 */
function parseRetryAfterHeader(errorBody: any): number | undefined {
  if (!errorBody) return undefined;

  // Check for retry_after in body
  if (typeof errorBody.retry_after === "number") {
    return errorBody.retry_after * 1000;
  }

  // Check for retryAfter in body
  if (typeof errorBody.retryAfter === "number") {
    return errorBody.retryAfter * 1000;
  }

  return undefined;
}

/**
 * Create a ConvexError with embedded ApiError
 * This allows the frontend to extract and display the error properly
 */
export function createApiConvexError(apiError: ApiError): ConvexError<string> {
  // Serialize ApiError data into the ConvexError message for frontend extraction
  // The frontend uses extractApiError() from shared-types to reconstruct the ApiError
  const errorData = {
    type: "api_error" as const,
    errorCode: apiError.errorCode,
    provider: apiError.provider,
    category: apiError.category,
    severity: apiError.severity,
    userMessage: apiError.userMessage,
    retryable: apiError.retryable ?? false,
    suggestedAction: apiError.suggestedAction,
    actionLabel: apiError.actionLabel,
    actionUrl: apiError.actionUrl,
    retryAfterMs: apiError.retryAfterMs,
    technicalMessage: apiError.technicalMessage,
    originalStatus: apiError.originalStatus,
    timestamp: apiError.timestamp,
  };

  // Use JSON string to embed structured data in ConvexError
  // Frontend can parse this to reconstruct ApiError
  return new ConvexError(JSON.stringify(errorData));
}

/**
 * Check if an error is a user-actionable API error
 * (i.e., requires user to do something, not just wait/retry)
 */
export function isUserActionableApiError(apiError: ApiError): boolean {
  const userActionableCategories: ApiErrorCategory[] = [
    "authentication",
    "authorization",
    "quota_exhausted",
    "rate_limit_exceeded",
  ];
  return userActionableCategories.includes(apiError.category);
}

/**
 * Check if error should block the pipeline
 * Based on user preference: block immediately on quota/credit errors
 */
export function shouldBlockPipeline(apiError: ApiError): boolean {
  const blockingCategories: ApiErrorCategory[] = [
    "authentication",
    "authorization",
    "quota_exhausted",
    "subscription_paused",
    "rate_limit_exceeded",
  ];
  return blockingCategories.includes(apiError.category);
}

// Re-export types and codes for convenience
export type { ApiError, ApiProvider, ApiErrorCategory, ErrorSeverity, ErrorAction };
export { API_ERROR_CODES };
